crates/wasi/src/snapshots/preview_1.rs

```diff
@@ -0,0 +1,513 @@
+
+    RawWaker::new(std::ptr::null::<()>(), vtable)
+}
+
+// Creates a dummy waker which does *nothing*, as the future itsef polls to ready at first poll
+// A waker is needed to do any polling at all, as it is the primary constituent of the `Context` for polling
+fn run_in_dummy_executor<F: std::future::Future>(f: F) -> Result<F::Output, Trap> {
+    let mut f = Pin::from(Box::new(f));
+    let waker = unsafe { Waker::from_raw(dummy_raw_waker()) };
+    let mut cx = Context::from_waker(&waker);
+    match f.as_mut().poll(&mut cx) {
+        std::task::Poll::Ready(val) => Ok(val),
+        std::task::Poll::Pending => Err(Trap::new("Cannot wait on pending future")),
+    }
+}
+
+// Creates the function item `add_wasi_snapshot_preview1_to_wasmi_linker` which when called adds all
+// `wasi preview_1` functions to the linker
+macro_rules! impl_add_to_linker_for_funcs {
+    (
+        $(
+            $( #[$docs:meta] )*
+            fn $fname:ident ($( $arg:ident : $typ:ty ),* $(,)? ) -> $ret:tt
+        );+ $(;)?
+    ) => {
+        fn add_wasi_snapshot_to_wasmi_linker<'a, T, U>(
+            linker: &mut Linker<T>,
+            mut store_ctx: impl AsContextMut<UserState = T>,
+            wasi_ctx: impl Fn(&mut T) -> &mut U + Send + Sync + Copy + 'static)
+        -> Result<(), Error>
+        where U: wasi_common::snapshots::preview_1::wasi_snapshot_preview1::WasiSnapshotPreview1 +
+                 wasi_common::snapshots::preview_1::wasi_snapshot_preview1::UserErrorConversion
+        {
+            let mut store = store_ctx.as_context_mut();
+            $(
+                $(#[$docs])*
+                linker.define(
+                    "wasi_snapshot_preview1",
+                    stringify!($fname),
+                    Func::wrap(&mut store, move|mut caller: Caller<'_, T>, $($arg : $typ,)*| {
+                        let result = async {
+                            let mem = match  caller.get_export("memory") {
+                                Some(Extern::Memory(m)) => m,
+                                _ => return Err(Trap::new("missing required memory export".to_string())),
+                            };
+                            let(mem, ctx) = mem.data_and_store_mut(&mut caller);
+                            let ctx = wasi_ctx(ctx);
+                            let mem = wiggle::wasmtime::WasmtimeGuestMemory::new(mem);
+
+                            match wasi_common::snapshots::preview_1::wasi_snapshot_preview1::$fname(ctx, &mem, $($arg,)*).await {
+                                Ok(r) => Ok(<$ret>::from(r)),
+                                Err(wiggle::Trap::String(err)) => Err(Trap::new(err)),
+                                Err(wiggle::Trap::I32Exit(i)) => Err(Trap::i32_exit(i)),
+
+                            }
+                        };
+                        run_in_dummy_executor(result)?
+                    })
+                )?;
+
+            )+
+    Ok(())
+    }
+}
+}
+
+impl_add_to_linker_for_funcs!(
```

Comment:
Wow this macro now looks very clean, thanks a lot for the improvements!

Reply:
Sure!